Update ICRC-28 and 34 (#191)
dfinity/wg-identity-authentication: topics/icrc_34_get_delegation.md

# ICRC-34: Delegation

[![Status Badge](https://img.shields.io/badge/STATUS-DRAFT-ffcc00.svg)](https://github.com/orgs/dfinity/projects/31)
[![Extension Badge](https://img.shields.io/badge/EXTENDS-ICRC--25-ffcc222.svg)](./icrc_25_signer_interaction_standard.md)
[![Standard Issue](https://img.shields.io/badge/ISSUE-ICRC--34-blue?logo=github)](https://github.com/dfinity/wg-identity-authentication/issues/115)

<!-- TOC -->
* [ICRC-34: Delegation](#icrc-34-delegation)
  * [Summary](#summary)
  * [Method](#method)
  * [Scope (according to the ICRC-25 standard)](#scope-according-to-the-icrc-25-standard)
    * [Example RPC Permission Request](#example-rpc-permission-request)
  * [icrc-25-supported-standards](#icrc-25-supported-standards)
  * [Request params](#request-params)
    * [Example RPC Delegation Request](#example-rpc-delegation-request)
  * [Result](#result)
    * [Example RPC Delegation Response](#example-rpc-delegation-response)
  * [Message Processing](#message-processing)
  * [Errors](#errors)
<!-- TOC -->

## Summary

> When a relying party wants to remove wallet approval prompts and sign on behalf of its users, it uses the icrc34_delegation 
> method described below. This method generates a delegation chain, an object with a stable identifier, allowing the relying 
> party to sign transactions on behalf of the user.

Reviewer: I would rephrase this as `avoid going through the signer and user approval for each individual call`.
Also the fact that the relying party does not need to stay connected to the signer when it has a delegation is an interesting aspect that might be worth mentioning.

```suggestion
When a relying party wants to avoid going through the signer and user approval for each individual call, it can use the icrc34_delegation method described below. 

This method generates a delegation chain, an object with a stable identifier, allowing the relying party to sign transactions on behalf of the user without having to go through the signer and user approval. This also means the signer does not need to stay connected to the signer to make calls on behalf of the user.
```

> The delegation chain's stable identifier can take two forms: an account that may include ICP, ICRC-1, and other assets 
> (Account Delegation), or an account designed exclusively for the relying party (Relying Party Delegation).

Reviewer: Since ICP is an ICRC-1 token, I would suggest removing ICP from the example.

Also would suggest below change with a numbered list to clearly list each type one by one.

```suggestion
The delegation chain's stable identifier can be either: 
1. **Account Delegation**: an identity that may hold tokens and other assets.
2. **Relying Party Delegation**: an identity designed exclusively for the relying party (Relying Party Delegation).
```

> 

Reviewer: The use case of a Relying Party Delegation is still unclear, maybe below explanation might help?

```suggestion

Example use cases of a Relying Party Delegation:
- Signer could give the user a choice to stay anonymous with a service.
- Differentiate between calls made with user approval (Account identifier) and without user approval (Relying Party Delegation identifier). Allowing for fine grained security levels per identifier.
- Exclusive identifier within the Relying Party platform to stay isolated from identifiers of other Relying Party platforms.

```

If a relying party wants to receive an Account Delegation, the `icrc34_delegation` request **MUST** include canisters it
controls as `targets` where each canister **MUST** implement the `icrc28_trusted_origins` endpoint as per the 
[ICRC-28](./icrc_28_trusted_origins.md) standard.

Relying Party Delegations **MUST** be restricted by the signer to individual relying parties.

> Signers **SHOULD** give users the choice to authenticate with their wallets (if available) or with an app-specific identifier.

Reviewer: I would personally consider keeping this out of the spec, a signer could decide to not give the user choice but decide based on the type of delegation requested. Which in most signers seems a more straightforward implementation.

Basically giving a user the choice to return a RP delegation when an Account delegation is requested is an advanced implementation to allow for anonymity but is also has technical challenges like what to return in the accounts request etc.

I can understand the reasons for adding it to the spec, but I would at least change it to `COULD` since the main goal here is probably to indicate that signers _could_ return a RP delegation when a Account delegation was requested.

```suggestion
Signers **COULD** give users the choice to authenticate with their relying party specific identifier instead of accounts (if available).
```

> Relying parties **SHOULD** remove targets if they want to isolate users from the rest of the ecosystem.

Reviewer: This could be one purpose, but shouldn't be the main purpose of Relying Party delegations. We obviously don't want to isolate users intentionally.

Also a dapp that uses Relying Party delegations could still request the list of accounts and communicate the principal from this list as user identifier across it's platform, making the user data thus non isolated from the rest of the ecosystem.

Also using a Relying Party delegation to isolate users would run into issues, since it's not an Account delegation for a principal returned in the list of accounts, it can't hold assets that can be managed by the wallet.

Instead I would recommend to use Relying Party delegations for more technical use cases, which I tried to suggest above in the summary. 

Nonetheless, above summary should attempt to clarify the difference between a Account and Relying Party delegation. This part of the spec should focus instead on clearly indicating how to get one type of delegation or the other. 

```suggestion
Relying parties **MUST** remove targets if they want to receive a Relying Party delegation.
```

Author: Good point.

Only thought is the difference between **must** and **should** in this context - if a relying party wants to receive an RP delegation, and they **could** receive one even if targets are included, does the clause here need to be "must"?

Reviewer: I would say yes it should be `MUST` since a RP delegation is not guaranteed when targets are defined but a side effect of the user deciding to be anonymous or targets not passing ICRC-28 validation.

So to be _guaranteed_ to receive a RP delegation, a RP must remove the targets.

Reviewer: Added guaranteed to the sentence since that should probably clarify it.
